src/core/utils/BinaryTree.scala

```diff
@@ -0,0 +1,29 @@
+package mutatus.utils
+
+sealed trait BinaryTree[+A] {
+  import scala.annotation.tailrec
+  import BinaryTree._
+  def value: Option[A] = this match {
+    case n: Node[A] => Some(n.v)
+    case l: Leaf[A] => Some(l.v)
+    case Empty      => None
+  }
+
+  def left: Option[BinaryTree[A]] = this match {
+    case n: Node[A] => Some(n.l)
+    case l: Leaf[A] => None
+    case Empty      => None
+  }
+
+  def right: Option[BinaryTree[A]] = this match {
+    case n: Node[A] => Some(n.r)
+    case l: Leaf[A] => None
+    case Empty      => None
+  }
+}
+
+object BinaryTree {
+  case class Node[A](v: A, l: BinaryTree[A], r: BinaryTree[A]) extends BinaryTree[A]
+  case class Leaf[A](v: A) extends BinaryTree[A]
```

Comment:
Very minor point, but I'd prefer `value`, `left` and `right` for the parameter names.

Reply:
Is there a reason why you're using pattern matching instead of inheritance for these methods, by the way?

Reply:
Having said all that, I see the issues with naming conflicts now. It seems like there might be an easier way to do this, but it's not too important.